File: kattis/guess.py
while True:
    try:        n = int(input())
    except:     break
    if not n:   break

    stack       = list()
    queue       = list()
    prioq       = list()
    li          = list()

    dstruct = [True, True, True]

    for _ in range(n):
        line    = list(map(int, input().split(" ")))
        if line[0] == 1:
            stack.append(line[1])
            queue.append(line[1])
            prioq.append(line[1])
            li.append(line[1])
        elif line[0] == 2:
            prioq.sort()
            if len(li) == 0:
                dstruct = [False, False, False]
                break
            li.pop()

            if dstruct[0] == True:
                if line[1] != stack[-1]:
                    dstruct[0] = False
                stack.pop()

            if dstruct[1] == True:
                if line[1] != queue[0]:
                    dstruct[1] = False
                queue.pop(0)

            if dstruct[2] == True:
                if line[1] != prioq[-1]:
                    dstruct[2] = False
                prioq.pop()

    if      dstruct == [False, False, False]:
        print("impossible")
    elif    dstruct == [True, False, False]:
        print("stack")
    elif    dstruct == [False, True, False]:
        print("queue")
    elif    dstruct == [False, False, True]:
        print("priority queue")
    else:
        print("not sure")

File: kattis/guess.go
package main

import (
  "fmt"
  "sort"
)


func main() {
  var l, i, n int
  for {
    fmt.Printf("INput")
    if _, err := fmt.Scan(&l); err != nil { return; }
    S := true; Q := true; P := true
    var queue, stack, sorted []int
    for {
      b := 0
      if l == 0 {
        break
      }
      l--
      if _, err := fmt.Scan(&i, &n); err != nil {
        break
      }
      if i == 1 {
        queue = append(queue, n)
        stack = append(stack, n)
        sorted = append(sorted, n)
      } else {
        il := len(stack) - 1
        if b > il {
          S, Q, P = false, false, false
          break
        }
        b++
        sort.Ints(sorted)
        x := queue[0]
        y := stack[il];
        z := sorted[il];
        queue = queue[1:]
        stack = stack[:il]
        sorted = sorted[:il]
        if x != n { Q = false; }
        if y != n { S = false; }
        if z != n { P = false; }
      }
    }

    if Q && P { fmt.Println("not sure");
    } else if S && P { fmt.Println("not sure");
    } else if S && Q { fmt.Println("impossible");
    } else if Q { fmt.Println("queue");
    } else if S { fmt.Println("stack");
    } else if P { fmt.Println("priority queue")
    } else { fmt.Println("impossible"); }
  }
}
